### PR TITLE
Prevent Duplicates at the Application Level with Additional Validation and Unique Index

### LINKED ISSUE #47: Create Contacts BE model unique name fix
When creating a contact, users could accidentally create duplicate contacts due to:
	1.	Formatting Inconsistencies:
	•	Contacts with identical names were treated as different due to differences in case (e.g., John vs. john) or trailing/leading whitespaces (e.g., John vs. John).
	•	While the Contact model included a uniqueness validation scoped to user_id and last_name, it lacked robustness against case and whitespace inconsistencies.
	2.	Concurrency and Race Conditions:
	•	Without a database-level unique index, simultaneous requests to create the same contact for the same user could bypass application-level validations, leading to duplicate records in the database.

Solution:
	1.	Application-Level Fixes:
	•	Introduced a before_validation callback to normalize first_name and last_name by trimming whitespace and capitalizing names.
	•	Ensured the uniqueness validation is scoped to user_id, first_name, and last_name, accounting for formatting inconsistencies.
	2.	Database-Level Fixes:
	•	Added a unique index to the contacts table for user_id, first_name, and last_name to enforce constraints at the database level. This prevents duplicates even under concurrent requests.

### PR DESCRIPTION
**Pull request by CCirbo**
## Type of Change
- [x] feature ⛲
- [x] bug fix 🐛
- [x] testing 🧪

## Description

This PR addresses duplicate Contact records at both the application and database levels. At the application level, a uniqueness validation has been added to first_name, scoped to last_name and user_id, ensuring that users cannot create duplicate contacts even with differences in case or whitespace.

A before_validation callback normalizes first_name and last_name by trimming whitespace and capitalizing them, further reducing the chances of duplicates due to formatting inconsistencies.

Additionally, a unique index has been added to the database to enforce this validation at the database level, ensuring data integrity and preventing race conditions that could lead to duplicate records.

Changes
	1.	Model Updates:
	•	Added uniqueness validation to first_name scoped to last_name and user_id, with a custom error message.
	•	Implemented a before_validation callback to normalize first_name and last_name.
	•	Updated validations for phone number and email formats.
	2.	Database Updates:
	•	Added a unique database index to the contacts table for first_name, last_name, and user_id to enforce uniqueness at the database level.
	3.	Test Coverage:
	•	Added tests to ensure names are capitalized and stripped of extra whitespace before saving.
	•	Added tests to confirm that duplicate contacts are invalid, even with differences in case or whitespace.

Future Considerations
While these changes address duplicates at both the application and database levels, further improvements could include:
	•	Enforcing additional constraints across other models that rely on uniqueness validations.

## Motivation and Context
Problem
When creating a contact, users could accidentally create duplicate contacts due to:
	1.	Formatting Inconsistencies:
	•	Contacts with identical names were treated as different due to differences in case (e.g., Jane vs. jane) or trailing/leading whitespaces (e.g., Jane vs. Jane).
	•	While the Contact model included a uniqueness validation scoped to user_id and last_name, it lacked robustness against case and whitespace inconsistencies.
	2.	Concurrency and Race Conditions:
	•	Without a database-level unique index, simultaneous requests to create the same contact for the same user could bypass application-level validations, leading to duplicate records in the database.

Solution
This PR resolves these issues by implementing a two-layer safeguard:
	1.	Application-Level Fixes:
	•	Introduced a before_validation callback to normalize first_name and last_name by trimming whitespace and capitalizing names.
	•	Ensured the uniqueness validation is scoped to user_id, first_name, and last_name, accounting for formatting inconsistencies.
	2.	Database-Level Fixes:
	•	Added a unique index to the contacts table for user_id, first_name, and last_name to enforce constraints at the database level. This prevents duplicates even under concurrent requests.

closes #47 
## Related Tickets
No related tickets however since the scope of this PR is only to fix unique name validations, no changes to any other tests or changes to update admin policies are in this PR. This PR is only for the contacts model unique name fix.

## Added Test?
- [x] Yes 🫡
 Unit/Model 
Two new tests for checking duplicate contacts not made if trailing whitespaces or case insensitivity.
All contact model tests are passing.

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: app/models/contact.rb [hunk 1]
```diff
@@ -5,9 +5,23 @@ class Contact < ApplicationRecord
...
+
   validates :first_name, presence: {message: "can't be blank"} 
   validates :last_name, presence: {message: "can't be blank"}
-  validates :first_name, uniqueness: { scope: [:last_name, :user_id], message: "and Last name already exist for this user" }
+  
+  validates :first_name, uniqueness: { 
+    scope: [:last_name, :user_id], 
+    case_sensitive: false,
+    message: "and Last name already exist for this user"
+}
```
Comment (reneemes) [inline app/models/contact.rb:17]: I'm so glad you caught this issue in the code and took the time to correct it. Excellent job adding more to this validation in the Contacts model. These changes are very beneficial in keeping our data clean and the application overall more user-friendly.

File: spec/models/contact_spec.rb [hunk 2]
```diff
@@ -38,6 +38,20 @@
...
+      end
+
+      it "does not allow duplicates with case or whitespace differences" do
+        Contact.create!(first_name: "Sandy", last_name: "Johnson", user: @user1)
+        duplicate_contact = Contact.new(first_name: "sandy", last_name: " johnson ", user: @user1)
+
+        expect(duplicate_contact).not_to be_valid
+        expect(duplicate_contact.errors.full_messages).to include("First name and Last name already exist for this user")
+      end
+
```
Comment (reneemes) [inline spec/models/contact_spec.rb:54]: I'm glad to see you tested this as well. Excellent job handling those edge cases.

**Review by litobot — APPROVED** [review]
The changes you've made along with the tests you've written to confirm your functionality greatly improves the UX by "idiot-proofing" the contact creation process.  Using `before_validation` to normalize names preventing duplicates in the database averts confusion and frustration for the User.

Using scope to build off existing validations is a great way to strengthen this process.

I like how you incorporated add_index to quickly verify whether or not the records already existed.
